### PR TITLE
Armor datums

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
New PR to replace #34173 since it cannot be reopened.

Basically what this does is replaces 90000+ lists on objects with around 139 shared armor datums of which new ones will be created and shared with future user objects.

:cl: Naksu
code: Slightly refactored object armor code
/:cl:

TODO:
- [x] VV support
- [x] guards near .vars access?

### CODE REVIEW

File: code/datums/armor.dm [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+/datum/armor/proc/modifyRating(melee = 0, bullet = 0, laser = 0, energy = 0, bomb = 0, bio = 0, rad = 0, fire = 0, acid = 0)
+  return getArmor(src.melee+melee, src.bullet+bullet, src.laser+laser, src.energy+energy, src.bomb+bomb, src.bio+bio, src.rad+rad, src.fire+fire, src.acid+acid)
+
+/datum/armor/proc/modifyAllRatings(modifier = 0)
+  return getArmor(src.melee+modifier, src.bullet+modifier, src.laser+modifier, src.energy+modifier, src.bomb+modifier, src.bio+modifier, src.rad+modifier, src.fire+modifier, src.acid+modifier)
+
+/datum/armor/proc/setRating(melee, bullet, laser, energy, bomb, bio, rad, fire, acid)
+  return getArmor((isnull(melee) ? src.melee : melee),\
+                  (isnull(melee) ? src.bullet : bullet),\
+                  (isnull(melee) ? src.laser : laser),\
```
Comment (XDTM) [inline code/datums/armor.dm]: isnull(laser) and so on, i suppose?

File: code/datums/armor.dm [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  src.melee = melee
+  src.bullet = bullet
+  src.laser = laser
+  src.energy = energy
+  src.bomb = bomb
+  src.bio = bio
+  src.rad = rad
+  src.fire = fire
+  src.acid = acid
+  GLOB.armorobjects[ARMORLISTID] = src
```
Comment (XDTM) [inline code/datums/armor.dm]: does this get updated when the armor is modified?

File: code/datums/armor.dm [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+/datum/armor/proc/getList()
+  return list("melee" = melee, "bullet" = bullet, "laser" = laser, "energy" = energy, "bomb" = bomb, "bio" = bio, "rad" = rad, "fire" = fire, "acid" = acid)
+
+/datum/armor/proc/attachArmor(datum/armor/AA)
+  return getArmor(src.melee+AA.melee, src.bullet+AA.bullet, src.laser+AA.laser, src.energy+AA.energy, src.bomb+AA.bomb, src.bio+AA.bio, src.rad+AA.rad, src.fire+AA.fire, src.acid+AA.acid)
+
+/datum/armor/proc/detachArmor(datum/armor/AA)
+  return getArmor(src.melee-AA.melee, src.bullet-AA.bullet, src.laser-AA.laser, src.energy-AA.energy, src.bomb-AA.bomb, src.bio-AA.bio, src.rad-AA.rad, src.fire-AA.fire, src.acid-AA.acid)
+
+/datum/armor/vv_edit_var(var_name, var_value)
```
Comment (lzimann) [inline code/datums/armor.dm]: i feel like the datum should just block var edits entirely and force people to use the dropdown, or at least require a confirmation that it's indeed what the person wanted

**Comment by vuonojenmustaturska** [comment]
image
image

oh right forgot to upload these.

File: code/datums/armor.dm (hunk 3 above)
Comment (vuonojenmustaturska) [reply code/datums/armor.dm]: yeah, i'm just not sure if vv_edit_var is the right place to present confirmation dialogs

Comment (vuonojenmustaturska) [reply code/datums/armor.dm]: added a check to /client/proc/modify_variables, same place where the pixel movement warnings are.

File: code/datums/armor.dm [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                  (isnull(laser) ? src.laser : laser),\
+                  (isnull(energy) ? src.energy : energy),\
+                  (isnull(bomb) ? src.bomb : bomb),\
+                  (isnull(bio) ? src.bio : bio),\
+                  (isnull(rad) ? src.rad : rad),\
+                  (isnull(fire) ? src.fire : fire),\
+                  (isnull(acid) ? src.acid : acid))
+
+/datum/armor/proc/getRating(rating)
+  return src.vars[rating]
```
Comment (vuonojenmustaturska) [inline code/datums/armor.dm]: wonder if a switch() would be more perfomant, .vars is actually slower than direct var access

File: code/datums/armor.dm [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+/datum/armor/proc/getList()
+  return list("melee" = melee, "bullet" = bullet, "laser" = laser, "energy" = energy, "bomb" = bomb, "bio" = bio, "rad" = rad, "fire" = fire, "acid" = acid)
+
+/datum/armor/proc/attachArmor(datum/armor/AA)
+  return getArmor(src.melee+AA.melee, src.bullet+AA.bullet, src.laser+AA.laser, src.energy+AA.energy, src.bomb+AA.bomb, src.bio+AA.bio, src.rad+AA.rad, src.fire+AA.fire, src.acid+AA.acid)
```
Comment (ninjanomnom) [inline code/datums/armor.dm]: `src.` is pointless here

File: code/datums/armor.dm [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+/datum/armor/proc/detachArmor(datum/armor/AA)
+  return getArmor(src.melee-AA.melee, src.bullet-AA.bullet, src.laser-AA.laser, src.energy-AA.energy, src.bomb-AA.bomb, src.bio-AA.bio, src.rad-AA.rad, src.fire-AA.fire, src.acid-AA.acid)
```
Comment (ninjanomnom) [inline code/datums/armor.dm]: same as above

File: code/datums/browser.dm [hunk 7]
```diff
@@ -223,19 +223,25 @@
 /datum/browser/modal/listpicker
 	var/valueslist = list()
 
-/datum/browser/modal/listpicker/New(User,Message,Title,Button1="Ok",Button2,Button3,StealFocus = 1, Timeout = FALSE,list/values,Radio=FALSE)
+/datum/browser/modal/listpicker/New(User,Message,Title,Button1="Ok",Button2,Button3,StealFocus = 1, Timeout = FALSE,list/values,Type="checkbox")
```
Comment (ninjanomnom) [inline code/datums/browser.dm]: please use a different arg name

File: code/datums/browser.dm [hunk 8]
```diff
@@ -266,14 +272,14 @@
 	opentime = 0
 	close()
 
-/proc/presentpicker(var/mob/User,Message, Title, Button1="Ok", Button2, Button3, StealFocus = 1,Timeout = 6000,list/values,Radio=FALSE)
+/proc/presentpicker(var/mob/User,Message, Title, Button1="Ok", Button2, Button3, StealFocus = 1,Timeout = 6000,list/values, Type = "checkbox")
```
Comment (ninjanomnom) [inline code/datums/browser.dm]: same as above

File: code/datums/browser.dm [hunk 9]
```diff
@@ -266,14 +272,14 @@
...
-/proc/presentpicker(var/mob/User,Message, Title, Button1="Ok", Button2, Button3, StealFocus = 1,Timeout = 6000,list/values,Radio=FALSE)
+/proc/presentpicker(var/mob/User,Message, Title, Button1="Ok", Button2, Button3, StealFocus = 1,Timeout = 6000,list/values, Type = "checkbox")
 	if (!istype(User))
 		if (istype(User, /client/))
 			var/client/C = User
 			User = C.mob
 		else
 			return
-	var/datum/browser/modal/listpicker/A = new(User, Message, Title, Button1, Button2, Button3, StealFocus,Timeout, values,Radio)
+	var/datum/browser/modal/listpicker/A = new(User, Message, Title, Button1, Button2, Button3, StealFocus,Timeout, values, Type)
```
Comment (ninjanomnom) [inline code/datums/browser.dm]: same as above

**Review by ninjanomnom — DISMISSED** [review]

**Comment by PKPenguin321** [comment]
This looks great, the UI especially

File: code/datums/armor.dm (hunk 4 above)
Comment (Cyberboss) [reply code/datums/armor.dm]: ehhhhh it should be fine. I don't think damage calculation is a real performance nerce

**Comment by outbools** [comment]
armor component nao

**Comment by Bawhoppen** [comment]
Be warned, as oranges may claim this is a sec buff

**Comment by vuonojenmustaturska** [comment]
@outbools that'd mean adding 90k+ datums, essentially throwing away any gains this idea has.

File: code/datums/armor.dm [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  src.bullet = bullet
+  src.laser = laser
+  src.energy = energy
+  src.bomb = bomb
+  src.bio = bio
+  src.rad = rad
+  src.fire = fire
+  src.acid = acid
+  tag = ARMORID
+  datum_flags = DF_USE_TAG
```
Comment (Cyberboss) [inline code/datums/armor.dm]: this should be set in the definition

File: code/datums/armor.dm [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+/datum/armor/proc/modifyRating(melee = 0, bullet = 0, laser = 0, energy = 0, bomb = 0, bio = 0, rad = 0, fire = 0, acid = 0)
+  return getArmor(src.melee+melee, src.bullet+bullet, src.laser+laser, src.energy+energy, src.bomb+bomb, src.bio+bio, src.rad+rad, src.fire+fire, src.acid+acid)
+
+/datum/armor/proc/modifyAllRatings(modifier = 0)
+  return getArmor(src.melee+modifier, src.bullet+modifier, src.laser+modifier, src.energy+modifier, src.bomb+modifier, src.bio+modifier, src.rad+modifier, src.fire+modifier, src.acid+modifier)
```
Comment (Cyberboss) [inline code/datums/armor.dm]: more src

File: code/datums/armor.dm [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                  (isnull(laser) ? src.laser : laser),\
+                  (isnull(energy) ? src.energy : energy),\
+                  (isnull(bomb) ? src.bomb : bomb),\
+                  (isnull(bio) ? src.bio : bio),\
+                  (isnull(rad) ? src.rad : rad),\
+                  (isnull(fire) ? src.fire : fire),\
+                  (isnull(acid) ? src.acid : acid))
+
+/datum/armor/proc/getRating(rating)
+  return src.vars[rating]
```
Comment (Cyberboss) [inline code/datums/armor.dm]: src

File: code/datums/armor.dm [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+/datum/armor/proc/attachArmor(datum/armor/AA)
+  return getArmor(melee+AA.melee, bullet+AA.bullet, laser+AA.laser, energy+AA.energy, bomb+AA.bomb, bio+AA.bio, rad+AA.rad, fire+AA.fire, acid+AA.acid)
+
+/datum/armor/proc/detachArmor(datum/armor/AA)
+  return getArmor(melee-AA.melee, bullet-AA.bullet, laser-AA.laser, energy-AA.energy, bomb-AA.bomb, bio-AA.bio, rad-AA.rad, fire-AA.fire, acid-AA.acid)
+
+/datum/armor/vv_edit_var(var_name, var_value)
+  . = ..()
+  src.tag = ARMORID
```
Comment (Cyberboss) [inline code/datums/armor.dm]: [b]rc

Comment (Cyberboss) [reply code/datums/armor.dm]: Actually if they try to change tag you should reject the edit so they understand it can't be

Comment (Cyberboss) [reply code/datums/armor.dm]: via returning FALSE and not calling ..()

File: code/datums/browser.dm [hunk 14]
```diff
@@ -223,19 +223,25 @@
 /datum/browser/modal/listpicker
 	var/valueslist = list()
 
-/datum/browser/modal/listpicker/New(User,Message,Title,Button1="Ok",Button2,Button3,StealFocus = 1, Timeout = FALSE,list/values,Radio=FALSE)
+/datum/browser/modal/listpicker/New(User,Message,Title,Button1="Ok",Button2,Button3,StealFocus = 1, Timeout = FALSE,list/values,inputtype="checkbox")
 	if (!User)
 		return
 
 	var/output =  {"<form><input type="hidden" name="src" value="[REF(src)]"><ul class="sparse">"}
```
Comment (Cyberboss) [inline code/datums/browser.dm]: make this a list

File: code/datums/datumvars.dm [hunk 15]
```diff
@@ -796,6 +796,41 @@
 				return
 			offer_control(M)
 
+		else if (href_list["modarmor"])
+			if(!check_rights(0))
```
Comment (Cyberboss) [inline code/datums/datumvars.dm]: NONE

File: code/game/objects/objs.dm [hunk 16]
```diff
@@ -37,8 +37,11 @@
 
 /obj/Initialize()
 	. = ..()
-	if (!armor)
-		armor = list(melee = 0, bullet = 0, laser = 0, energy = 0, bomb = 0, bio = 0, rad = 0, fire = 0, acid = 0)
+	if (islist(armor))
+		armor = getArmor(arglist(armor))
+	else if (!istype(armor, /datum/armor))
+		armor = getArmor()
```
Comment (Cyberboss) [inline code/game/objects/objs.dm]: stack_trace here since it's an error

File: code/datums/armor.dm [hunk 17]
```diff
@@ -59,7 +59,9 @@
   return getArmor(melee-AA.melee, bullet-AA.bullet, laser-AA.laser, energy-AA.energy, bomb-AA.bomb, bio-AA.bio, rad-AA.rad, fire-AA.fire, acid-AA.acid)
 
 /datum/armor/vv_edit_var(var_name, var_value)
+  if (var_name == "tag")
```
Comment (Cyberboss) [inline code/datums/armor.dm]: "tag" => NAMEOF(src, tag)

File: code/datums/armor.dm [hunk 18]
```diff
@@ -59,7 +59,9 @@
...
+    return FALSE
   . = ..()
-  src.tag = ARMORID
+  src.tag = ARMORID // update tag in case armor values were edited
```
Comment (Cyberboss) [inline code/datums/armor.dm]: srrcccc

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
>Tfw github tells you to review out of date shit

File: code/datums/armor.dm [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+  var/melee
+  var/bullet
+  var/laser
+  var/energy
+  var/bomb
+  var/bio
+  var/rad
+  var/fire
+  var/acid
+  datum_flags = DF_USE_TAG
```
Comment (Cyberboss) [inline code/datums/armor.dm]: Move this up to the top please, I try to enforce new vars coming last followed by statics

File: code/datums/armor.dm [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+  return getArmor(melee+AA.melee, bullet+AA.bullet, laser+AA.laser, energy+AA.energy, bomb+AA.bomb, bio+AA.bio, rad+AA.rad, fire+AA.fire, acid+AA.acid)
+
+/datum/armor/proc/detachArmor(datum/armor/AA)
+  return getArmor(melee-AA.melee, bullet-AA.bullet, laser-AA.laser, energy-AA.energy, bomb-AA.bomb, bio-AA.bio, rad-AA.rad, fire-AA.fire, acid-AA.acid)
+
+/datum/armor/vv_edit_var(var_name, var_value)
+  if (var_name == "tag")
+    return FALSE
+  . = ..()
+  src.tag = ARMORID // update tag in case armor values were edited
```
Comment (Cyberboss) [inline code/datums/armor.dm]: src

File: code/datums/armor.dm [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+  return list("melee" = melee, "bullet" = bullet, "laser" = laser, "energy" = energy, "bomb" = bomb, "bio" = bio, "rad" = rad, "fire" = fire, "acid" = acid)
+
+/datum/armor/proc/attachArmor(datum/armor/AA)
+  return getArmor(melee+AA.melee, bullet+AA.bullet, laser+AA.laser, energy+AA.energy, bomb+AA.bomb, bio+AA.bio, rad+AA.rad, fire+AA.fire, acid+AA.acid)
+
+/datum/armor/proc/detachArmor(datum/armor/AA)
+  return getArmor(melee-AA.melee, bullet-AA.bullet, laser-AA.laser, energy-AA.energy, bomb-AA.bomb, bio-AA.bio, rad-AA.rad, fire-AA.fire, acid-AA.acid)
+
+/datum/armor/vv_edit_var(var_name, var_value)
+  if (var_name == "tag")
```
Comment (Cyberboss) [inline code/datums/armor.dm]: "tag" => NAMEOF(src, tag)

File: code/datums/datumvars.dm [hunk 22]
```diff
@@ -796,6 +796,41 @@
...
+			                  bullet = text2num(result["values"]["bullet"]),\
+			                  laser = text2num(result["values"]["laser"]),\
+			                  energy = text2num(result["values"]["energy"]),\
+			                  bomb = text2num(result["values"]["bomb"]),\
+			                  bio = text2num(result["values"]["bio"]),\
+			                  rad = text2num(result["values"]["rad"]),\
+			                  fire = text2num(result["values"]["fire"]),\
+			                  acid = text2num(result["values"]["acid"]))
+				log_admin("[key_name(usr)] modified the armor on [O] ([O.type]) to melee: [O.armor.melee], bullet: [O.armor.bullet], laser: [O.armor.laser], energy: [O.armor.energy], bomb: [O.armor.bomb], bio: [O.armor.bio], rad: [O.armor.rad], fire: [O.armor.fire], acid: [O.armor.acid]")
+				message_admins("<span class='notice'>[key_name(usr)] modified the armor on [O] ([O.type]) to melee: [O.armor.melee], bullet: [O.armor.bullet], laser: [O.armor.laser], energy: [O.armor.energy], bomb: [O.armor.bomb], bio: [O.armor.bio], rad: [O.armor.rad], fire: [O.armor.fire], acid: [O.armor.acid]</span>")
```
Comment (Cyberboss) [inline code/datums/datumvars.dm]: key_name_admin in message_admins

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
Seems functionally fine to me

File: code/datums/armor.dm (hunk 21 above)
Comment (vuonojenmustaturska) [reply code/datums/armor.dm]: I'll be honest, i don't really see the point?

Comment (Cyberboss) [reply code/datums/armor.dm]: It's for compile time checking of the var name. I know `tag` is probably never going to change but ODR et al.

**Review by Cyberboss — DISMISSED** [review]

**Comment by ShizCalev** [comment]
#34928

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**